bpo-24882: Let ThreadPoolExecutor reuse idle threads before creating new thread

iUnknwn opened the pull request:
The change fixes an issue where ThreadPoolExecutor doesn't check if existing threads are idle before spinning up new ones. To do this, we use a simple counter within the Executor to track how many threads are idle, and atomically increment and decrement the count as needed. 

One question - the previous code to spin up a new thread in `_adjust_thread_count` does not appear thread safe - if two threads are both submitting items to the executor at the same time, the call to check the number of threads could be invalid. Is this something that should also be fixed, or is this not an issue because of the global interpreter lock?


<!-- issue-number: [bpo-24882](https://bugs.python.org/issue24882) -->
https://bugs.python.org/issue24882
<!-- /issue-number -->

iUnknwn committed:
Fixes issue 24882

iUnknwn committed:
Add news file entry for change.

iUnknwn committed:
Change test_concurrent_futures.ThreadPoolShutdownTest

Adjust the shutdown test so that, after submitting three jobs
to the executor, the test checks for less than three threads,
instead of looking for exactly three threads.

If idle threads are being recycled properly, then we should have
less than three threads.

pitrou reviewed (commented):
As discussed on the issue tracker, I don't think it's a good idea to add this.

MojoVampire commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -318,7 +318,7 @@ def test_threads_terminate(self):
         self.executor.submit(mul, 6, 7)
         self.executor.submit(mul, 3, 14)
-        self.assertEqual(len(self.executor._threads), 3)
+        self.assertTrue(len(self.executor._threads) < 3)
```
You've submitted three things, so why wouldn't it be possible for it to have launched three threads? Granted, the work per task is cheap, so it's possible (maybe even likely, given GIL interference) that the first task is done in time for its thread to be reused by the third task, but that's not guaranteed, is it?

tomMoral replied:
I agree that this test might fail randomly. It would be better to force the collection of the 3 tasks with calls to `result` before launching the next one and verify that only one thread has been used.

tomMoral commented on Lib/concurrent/futures/thread.py:
```diff
@@ -129,6 +136,8 @@ def __init__(self, max_workers=None, thread_name_prefix='',
         self._max_workers = max_workers
         self._work_queue = queue.SimpleQueue()
+        self._idle_lock = threading.Lock()
+        self._idle_count = 0
```
Why don't you use a `threading.Semaphore` for  this?
It makes the implementation easier to read as you don't need to handle the lock + increments but only rely on `acquire` and `release`.

tomMoral commented on Lib/concurrent/futures/thread.py:
```diff
@@ -155,12 +164,17 @@ def submit(self, fn, *args, **kwargs):
+        with self._idle_lock:
+            if self._idle_count > 0:
+                self._idle_count -= 1
+                return
```
If `_idle_count` is a semaphore, you simply need to use:
```
if self._idle_count.acquire(timeout=0):
    return
```

tomMoral commented on Lib/concurrent/futures/thread.py:
```diff
@@ -80,7 +80,14 @@ def _worker(executor_reference, work_queue, initializer, initargs):
+                # attempt to increment idle count
+                executor = executor_reference()
+                if executor is not None:
+                    executor._increase_idle_count()
```
If `_idle_count` is a semaphore, you just need to increase the number with `release`:
```
executor._idle_count.release()
```

tomMoral reviewed (commented):
The implementation seems to be doing the job. I think it would be a bit simpler by using a `Semaphore` instead of a lock protected counter.

I am unsure about the race conditions that might occur. Here, I do not see any obvious ones but to be on the safe side, it would be nice to have a test checking that you can indeed saturate the `ThreadPoolExecutor` with many small tasks. for instance:

```
def test_saturation(self):
    list(self.executor.map(mul, range(100 * self.executor._max_workers), y=0))
    self.assertEqual(len(self.executor._threads), self.executor._max_workers)
```

iUnknwn committed:
Switched idle count to semaphor, Updated tests

As suggested by reviewer tomMoral, swapped lock-protected counter
with a semaphore to track the number of unused threads.

Adjusted test_threads_terminate to wait for completiton of the
previous future before submitting a new one (and checking the
number of threads used).

Also added a new test to confirm the thread pool can be saturated.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -315,10 +315,10 @@ def _prime_executor(self):
-        self.assertEqual(len(self.executor._threads), 3)
+        self.executor.submit(mul, 21, 2).result()
+        self.executor.submit(mul, 6, 7).result()
+        self.executor.submit(mul, 3, 14).result()
```
I'd rather see the test fixed to test the proper condition (i.e. all 3 threads launched, then properly terminated).
For example (untested):
```
def acquire_lock(lock):
    lock.acquire()

def test_threads_terminate(self):
    sem = threading.Semaphore(0)
    for i in range(3):
        self.executor.submit(acquire_lock, sem)
    self.assertEqual(len(self.executor._threads), 3)
    for i in range(3):
        sem.release()
    # etc.
```

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -715,6 +715,10 @@ def test_default_workers(self):
         self.assertEqual(executor._max_workers,
                          (os.cpu_count() or 1) * 5)
 
+    def test_saturation(self):
```
See comment above for a proper way to do this.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -715,6 +715,10 @@ def test_default_workers(self):
+    def test_saturation(self):
+        list(self.executor.map(lambda x: mul(0, x), range(100 * self.executor._max_workers)))
+        self.assertEqual(len(self.executor._threads), self.executor._max_workers)
+
```
Can you add a test launching one task at a time and checking that the number of threads doesn't grow?

pitrou commented on Lib/concurrent/futures/thread.py:
```diff
@@ -155,12 +163,15 @@ def submit(self, fn, *args, **kwargs):
     submit.__doc__ = _base.Executor.submit.__doc__
 
     def _adjust_thread_count(self):
+        #if idle threads are available, don't spin new threads
```
Style: use a space after `#`.

pitrou reviewed (changes requested):
Thank you for the update. Here are a few more comments, mostly around tests.

iUnknwn committed:
Updates tests as requested by pitrou.

iUnknwn committed:
Correct minor whitespace error.

pitrou committed:
Make test_saturation faster

tzongw commented on Lib/concurrent/futures/thread.py:
```diff
@@ -80,7 +80,14 @@ def _worker(executor_reference, work_queue, initializer, initargs):
+                # attempt to increment idle count
+                executor = executor_reference()
+                if executor is not None:
+                    executor._idle_semaphore.release()
```
Is `_idle_semaphore` presenting idle threads? If more than `max_workers` jobs is submited at the same time, won't `_idle_semaphore` value greater than the thread count?
